query/src/lib.rs

```diff
@@ -8,4 +8,206 @@
+/// entity IDs. That'll get filtered out in the context of the
+/// database.
+pub enum PatternNonValuePlace {
+    Placeholder,
+    Variable(Variable),
+    Entid(u64),                       // Note unsigned.
+    Ident(NamespacedKeyword),
+}
+
+/// The `v` part of a pattern can be much broader: it can represent
+/// integers that aren't entity IDs (particularly negative integers),
+/// strings, and all the rest. We group those under `Constant`.
+pub enum PatternValuePlace {
+    Placeholder,
+    Variable(Variable),
+    EntidOrInteger(i64),
+    Ident(NamespacedKeyword),
+    Constant(NonIntegerConstant),
+}
+
+/*
+pub enum PullPattern {
+    Constant(Constant),
+    Variable(Variable),
+}
+
+pub struct Pull {
+    pub src: SrcVar,
+    pub var: Variable,
+    pub pattern: PullPattern,      // Constant, variable, or plain variable.
+}
+*/
+
+/*
+pub struct Aggregate {
+    pub fn_name: String,
+    pub args: Vec<FnArg>,
+}
+*/
+
+pub enum Element {
+    Variable(Variable),
+    // Aggregate(Aggregate),   // TODO
+    // Pull(Pull),             // TODO
+}
+
+/// A definition of the first part of a find query: the
+/// `[:find ?foo ?bar…]` bit.
+///
+/// There are four different kinds of find specs, allowing you to query for
+/// a single value, a collection of values from different entities, a single
+/// tuple (relation), or a collection of tuples.
+///
+/// Examples:
+///
+/// ```rust
+/// # extern crate edn;
+/// # extern crate mentat_query;
+/// # use edn::PlainSymbol;
+/// # use mentat_query::{Element, FindSpec, Variable};
+///
+/// # fn main() {
+///
+///   let elements = vec![
+///     Element::Variable(Variable(PlainSymbol("?foo".to_string()))),
+///     Element::Variable(Variable(PlainSymbol("?bar".to_string()))),
+///   ];
+///   let rel = FindSpec::FindRel(elements);
+///
+///   if let FindSpec::FindRel(elements) = rel {
+///     assert_eq!(2, elements.len());
+///   }
+///
+/// # }
+/// ```
+///
+pub enum FindSpec {
+    /// Returns an array of arrays.
+    FindRel(Vec<Element>),
+
+    /// Returns an array of scalars, usually homogeneous.
+    /// This is equivalent to mapping over the results of a `FindRel`,
+    /// returning the first value of each.
+    FindColl(Element),
+
+    /// Returns a single tuple: a heterogeneous array of scalars. Equivalent to
+    /// taking the first result from a `FindRel`.
+    FindTuple(Vec<Element>),
+
+    /// Returns a single scalar value. Equivalent to taking the first result
+    /// from a `FindColl`.
+    FindScalar(Element),
+}
+
+/// Returns true if the provided `FindSpec` returns at most one result.
+pub fn is_unit_limited(spec: &FindSpec) -> bool {
+    match spec {
+        &FindSpec::FindScalar(..) => true,
+        &FindSpec::FindTuple(..)  => true,
+        &FindSpec::FindRel(..)    => false,
+        &FindSpec::FindColl(..)   => false,
+    }
+}
+
+/// Returns true if the provided `FindSpec` cares about distinct results.
```

Comment:
Two of the find specs (scalar and tuple) produce only a single result. Those don't need to be run with `SELECT DISTINCT`, and that's this distinction.

There's a more broader definition of whether `DISTINCT` is needed: for example, in some kinds of accumulation or sampling projections we might not need to do it at the SQL level. This gives us the hook to add that.

I use the words "cares about" because `find` is generally defined in terms of producing distinct results at the Datalog level.